Sming/Libraries/OtaUpgrade/BasicOtaUpgradeStream.h

```diff
@@ -0,0 +1,181 @@
+ * and signature validation (if enabled) was successful, the updated slot is activated
+ * in the rBoot configuration.
+ * Call `hasError()` and/or check the public \c #errorCode member to determine if 
+ * everything went smoothly.
+ * 
+ * For further information on configuration options and the file format, 
+ * refer to the library's documentation.
+ * 
+ * @see `EncryptedOtaUpgradeStream` for encryption support.
+ */
+class BasicOtaUpgradeStream : public ReadWriteStream
+{
+public:
+	BasicOtaUpgradeStream();
+
+	/**
+	 * @brief Error code values
+	 */
+	enum ErrorCode {
+		NoError, ///< No error occured thus far (default value of \c #errorCode if `hasError()` returns false)
+		InvalidFormatError,   ///< Invalid/unsupported upgrade file format
+		UnsupportedDataError, ///< Some content of the upgrade file is not supported by this version of OtaUpgradeStream.
+		DecryptionError,	  ///< Decryption failed. Probably wrong decryption key.
+		NoRomFoundError, ///< The file did not contain a ROM image suitable for the start address of the slot to upgrade.
+		RomTooLargeError,   ///< The contained ROM image does not fit into the application firmware slot.
+		DowngradeError,		///< Attempt to downgrade to older firmware version.
+		VerificationError,  ///< Signature/checksum verification failed - updated ROM not activated
+		FlashWriteError,	///< Error while writing to Flash memory.
+		RomActivationError, ///< Error while activating updated ROM slot.
+		OutOfMemoryError,   ///< Dynamic memory allocation failed
+		InternalError		///< An unexpected error occured.
+	};
+
+	ErrorCode errorCode = NoError; ///< Error code. Only relevant if `hasError()` returns `true`.
+
+	/** @brief Convert error code to string.
+	 * @see #errorCode
+	 */
+	static String errorToString(ErrorCode code);
+
+	/** @brief Process chunk of upgrade file.
+	 * @param data Pointer to chunk of data.
+	 * @param size Size of chunk pointed to by \a data in bytes.
+	 * @return If less than \a size, an error occured. Check \c #errorCode for more details.
+	 * @note Even if `write()` never returns less than \a size it is not guaranteed that the upgrade was successful.
+	 * 		 Always use `hasError()` to determine success.
+	 */
+	size_t write(const uint8_t* data, size_t size) override;
+
+	/** @brief Returns true if an error happened during the upgrade process.
+	 * @see `errorCode`
+	 */
+	bool hasError() const
+	{
+		return (state == StateError);
+	}
+
+	// overrides from IDataSourceStream
+	uint16_t readMemoryBlock(char* data, int bufSize) override
+	{
+		return 0;
+	}
+	virtual int available() override
+	{
+		return 0;
+	}
+	bool isFinished() override
+	{
+		return true;
+	}
+
+protected:
+	void setError(ErrorCode ec);
+
+private:
+	/** Representation of the ROM slot to receive the new firmware */
+	struct Slot {
+		uint8_t index;
+		uint32_t address;
+		uint32_t size;
+		bool updated = false;
+		/** Determine the parameters of the slot to receive the upgrade. */
+		Slot();
+	} slot;
```

Comment:
Talking of packing, etc. you should re-arrange this structure so the elements are naturally aligned:

```
		uint32_t address;
		uint32_t size;
		uint8_t index;
		bool updated = false;
```

Reply:
In general I avoid distributing member variables. By convention, I always declare them after any member functions and together. One reason for that is so to try and maintain this 'natural alignment' for them.